uweb3/request.py

```diff
@@ -53,128 +79,395 @@ def _BaseCookie__set(self, key, real_value, coded_value):
+        self.request_payload = LimitedStream(env["wsgi.input"], self.content_length)
+        self._parse_functions = {
+            "application/json": self._parse_json,
+            "multipart/form-data": self._parse_multipart,
+            "application/x-www-form-urlencoded": self._parse_multipart,
+        }
+
+    def parse(self) -> IndexedFieldStorage:
+        # TODO: Handle streams
+        handler = self._parse_functions.get(self.mimetype, self._parse_multipart)
+        return handler()
+
+    def _parse_multipart(self) -> IndexedFieldStorage:
+        return IndexedFieldStorage(
+            io.BytesIO(self.request_payload.read(size=self.max_size)),
+            environ=self.env,
+            keep_blank_values=True,
+            limit=self.max_size,
+        )
+
+    def _parse_json(self) -> IndexedFieldStorage:
+        storage = IndexedFieldStorage()
+        try:
+            json_data = json.loads(self.request_payload.read(size=self.max_size))
+            storage.list = [
+                cgi.MiniFieldStorage(key, value) for key, value in json_data.items()
+            ]
+            return storage
+        except (json.JSONDecodeError, ValueError):
+            return storage
+
+
+class BaseRequest:
+    MAX_COOKIE_LENGTH: int = 4 * 1024  # 4KB
+    MAX_REQUEST_BODY_SIZE: int = 20 * 1024 * 1024  # 20MB
+
+    def __init__(self, env):
         self.env = env
-        self.headers = dict(self.headers_from_env(env))
-        self._out_headers = []
-        self._out_status = 200
-        self._response = None
         self.charset = "utf-8"
+        self.headers = dict(headers_from_env(env))
+        self.noparse = self.headers.get("accept", "").lower() == "application/json"
+
+        self.env["host"] = self.headers.get("Host", "").strip().lower()
+        self.env["REAL_REMOTE_ADDR"] = self._return_real_remote_addr()
+        self.env["mimetype"] = self._get_mimetype()
+
         self.method = self.env["REQUEST_METHOD"]
+        self.path = self.env["PATH_INFO"]
+
         self.vars = {
             "cookie": {
                 name: value.value
                 for name, value in Cookie(self.env.get("HTTP_COOKIE")).items()
             },
             "get": QueryArgsDict(parse_qs(self.env["QUERY_STRING"])),
+            "post": IndexedFieldStorage(),
+            "put": IndexedFieldStorage(),
+            "json": IndexedFieldStorage(),
+            "delete": IndexedFieldStorage(),
+            "files": IndexedFieldStorage(),
         }
-        self.env["host"] = self.headers.get("Host", "").strip().lower()
+
+    def _get_mimetype(self):
+        return self.env.get("CONTENT_TYPE", "").split(";")[0]
+
+    def _return_real_remote_addr(self):
+        """Returns the remote ip-address,
+        if there is a proxy involved it will take the last IP addres from the HTTP_X_FORWARDED_FOR list
+        """
+        try:
+            return self.env["HTTP_X_FORWARDED_FOR"].split(",")[-1].strip()
```

Comment:
We should probably make this header configurable, and only use it when we have some setting configured. Otherwise, when there is no proxy, an attacking user might set this header and spoof someone else's IP.